File: cadwiki-nuget/cadwiki.FileStore/Bitmaps.vb
Imports System
Imports System.Drawing.Imaging
Imports System.Drawing
Imports System.Windows.Media.Imaging
Imports System.Windows.Media
Imports PixelFormat = System.Drawing.Imaging.PixelFormat

Public Class Bitmaps

    Public Shared Function CreateBitmapSourceFromGdiBitmapForAutoCADButtonIcon(ByVal bitmap As Bitmap) As BitmapSource
        If bitmap Is Nothing Then Throw New ArgumentNullException("bitmap")

        Dim rect = New Rectangle(0, 0, bitmap.Width, bitmap.Height)

        Dim bitmapData = bitmap.LockBits(rect, ImageLockMode.ReadWrite, PixelFormat.Format32bppArgb)

        Try
            Dim size = rect.Width * rect.Height * 4

            Return BitmapSource.Create(bitmap.Width, bitmap.Height, bitmap.HorizontalResolution, bitmap.VerticalResolution, PixelFormats.Bgra32, Nothing, bitmapData.Scan0, size, bitmapData.Stride)
        Finally
            bitmap.UnlockBits(bitmapData)
        End Try
    End Function

End Class

File: cadwiki-nuget/cadwiki.FileStore/Bitmaps.cs
using System;
using System.Drawing;
using System.Drawing.Imaging;
using PixelFormat = System.Drawing.Imaging.PixelFormat;
using System.Windows.Media;
using System.Windows.Media.Imaging;

namespace cadwiki.FileStore
{

    public class Bitmaps
    {

        public static BitmapSource CreateBitmapSourceFromGdiBitmapForAutoCADButtonIcon(Bitmap bitmap)
        {
            if (bitmap is null)
            {
                return CreateDefaultAcadBitmap();
            }

            BitmapData bitmapData = null;
            try
            {
                var rect = new Rectangle(0, 0, bitmap.Width, bitmap.Height);
                bitmapData = bitmap.LockBits(rect, ImageLockMode.ReadWrite, PixelFormat.Format32bppArgb);
                int size = rect.Width * rect.Height * 4;
                return BitmapSource.Create(
                    bitmap.Width, 
                    bitmap.Height,
                    (double)bitmap.HorizontalResolution,
                    (double)bitmap.VerticalResolution,
                    PixelFormats.Bgra32,
                    null,
                    bitmapData.Scan0,
                    size,
                    bitmapData.Stride
                    );
            }
            catch (Exception)
            {
                return CreateDefaultAcadBitmap();
            }
            finally
            {
                if (bitmapData != null)
                {
                    bitmap.UnlockBits(bitmapData);
                }
            }
        }

        private static BitmapSource CreateDefaultAcadBitmap()
        {
            BitmapData bitmapData = null;
            Bitmap defaultBitmap = new Bitmap(100, 100);
            using (Graphics graphics = Graphics.FromImage(defaultBitmap))
            {
                graphics.Clear(System.Drawing.Color.White);
                graphics.DrawLine(Pens.Black, 0, 0, 100, 100);
                graphics.DrawLine(Pens.Black, 100, 0, 0, 100);
            }
            var rect = new Rectangle(0, 0, defaultBitmap.Width, defaultBitmap.Height);
            bitmapData = defaultBitmap.LockBits(rect, ImageLockMode.ReadWrite, PixelFormat.Format32bppArgb);
            int size = rect.Width * rect.Height * 4;

            return BitmapSource.Create(
                defaultBitmap.Width, 
                defaultBitmap.Height,
                (double)defaultBitmap.HorizontalResolution,
                (double)defaultBitmap.VerticalResolution,
                PixelFormats.Bgra32,
                null,
                bitmapData.Scan0,
                size,
                bitmapData.Stride
                );
        }
    }
}